external/VideoStreamer/src/VideoStreamer.cpp

```diff
@@ -0,0 +1,316 @@
+{
+    if (ppVideoStreamer == nullptr)
+    {
+        return E_POINTER;
+    }
+
+    VideoStreamerFFmpeg::InitFFmpeg();
+    auto pVS = new (std::nothrow) VideoStreamerFFmpeg();
+    if (pVS == NULL)
+    {
+        return E_OUTOFMEMORY;
+    }
+    return pVS->QueryInterface(__uuidof(IVideoStreamer), (void**)ppVideoStreamer);
+}
+
+// Create a new instance of the object.
+HRESULT CreateFFVideoStreamer(IVideoStreamer** ppVideoStreamer)
+{
+    return VideoStreamerFFmpeg::CreateInstance(ppVideoStreamer);
+}
+
+STDMETHODIMP VideoStreamerBase::QueryInterface(REFIID riid, void** ppv)
+{
+    static const QITAB qit[] =
+    {
+        QITABENT(VideoStreamerBase, IMFSampleGrabberSinkCallback),
+        QITABENT(VideoStreamerBase, IMFClockStateSink),
+        QITABENT(VideoStreamerBase, IVideoStreamer),
+        { 0 }
+    };
+    return QISearch(this, qit, riid, ppv);
+}
+
+STDMETHODIMP_(ULONG) VideoStreamerBase::AddRef()
+{
+    return InterlockedIncrement(&m_cRef);
+}
+
+STDMETHODIMP_(ULONG) VideoStreamerBase::Release()
+{
+    ULONG cRef = InterlockedDecrement(&m_cRef);
+    if (cRef == 0)
+    {
+        delete this;
+    }
+    return cRef;
+
+}
+
+// IMFClockStateSink methods.
+
+// In these example, the IMFClockStateSink methods do not perform any actions. 
+// You can use these methods to track the state of the sample grabber sink.
+
+STDMETHODIMP VideoStreamerBase::OnClockStart(MFTIME hnsSystemTime, LONGLONG llClockStartOffset)
+{
+    return S_OK;
+}
+
+STDMETHODIMP VideoStreamerBase::OnClockStop(MFTIME hnsSystemTime)
+{
+    return S_OK;
+}
+
+STDMETHODIMP VideoStreamerBase::OnClockPause(MFTIME hnsSystemTime)
+{
+    return S_OK;
+}
+
+STDMETHODIMP VideoStreamerBase::OnClockRestart(MFTIME hnsSystemTime)
+{
+    return S_OK;
+}
+
+STDMETHODIMP VideoStreamerBase::OnClockSetRate(MFTIME hnsSystemTime, float flRate)
+{
+    return S_OK;
+}
+
+// IMFSampleGrabberSink methods.
+
+STDMETHODIMP VideoStreamerBase::OnSetPresentationClock(IMFPresentationClock* pClock)
+{
+    return S_OK;
+}
+
+STDMETHODIMP VideoStreamerFFmpeg::OnProcessSample(REFGUID guidMajorMediaType, DWORD dwSampleFlags,
+    LONGLONG llSampleTime, LONGLONG llSampleDuration, const BYTE* pSampleBuffer,
+    DWORD dwSampleSize)
+{
+
+#ifdef TIGHT_LATENCY_CONTROL
+    auto tm = MFGetSystemTime();
+    auto delay = (tm - llSampleTime) / 10000;
+    INFOLOG2("\t SW Delay: %llu", delay);
+    if (delay > MAX_SINK_LATENCY)
+    {
+        std::cout << "\ndrop Sample :" << delay;
+        g_dropCount++;
+    }
+#endif
+    for (auto& avc : m_aAvfctx)
+    {
+        AVPacket pkt;
+        av_init_packet(&pkt);
+        pkt.data = (uint8_t*)pSampleBuffer;
+        pkt.size = dwSampleSize;
+        pkt.duration = llSampleDuration / 10000;
+        pkt.pts = llSampleTime / 10000;
+        pkt.dts = pkt.pts;
+        av_interleaved_write_frame(avc.second, &pkt);
+        av_packet_unref(&pkt);
+    }
+
+    return S_OK;
+}
+

```

Comment:
you are mixing hresults and exception code you have to have try catch statements like in the c++ coding guidelines

Reply:
The plan is to have all the code throwing and handle the catches in the application.
I shall be making it more uniform in further iterations

Reply:
BTW the ffmpeg functions are not supposed to throw exceptions on errors.